### DISCUSSION TITLE
Abstraction layer with UART to behave like a network connection when working with cellular modems.

### DISCUSSION THREAD

**Discussion by sethstenzel** [micropython/micropython#11039] [Ideas]
For a current work project, I'm using a cellular modem which I communicate with over uart. Most modems allow you to enter a raw mode where any data put on the wire is transmitted via whatever connection method was setup, udp, tcp, ssl, etc.

There are a few custom libraries I've had to create for doing some communication over TLS via these modems.
I've since decided I really need to create a general library that can be used so that these modem connections can be easily used by other micro python libraries that are expecting WLAN connections.

Before I start I'm reaching out here to see if anyone else has already done something similar or if anyone has and thoughts/pointers/advice.

Thanks.

**Comment by karfas** [comment]
> Most modems allow you to enter a raw mode where any data put on the wire is transmitted via whatever connection method was setup, udp, tcp, ssl, etc.
...
> There are a few custom libraries I've had to create for doing some communication over TLS via these modems.

Are you searching for a PPP or SLIP implementation ? 

While TCP/IP is not bound to a specific physical connection (you can use pigeons to send and receive packets), the usual protocols used to talk TCP/IP between serial endpoints are PPP or (older, slower) SLIP. 

I'm puzzled by your "TLS" implementation here. E.g. [RFC5246](https://tools.ietf.org/html/rfc5246) talks about "... At the lowest level, layered on top of some reliable transport protocol ... is the TLS Record Protocol ...". 

This has not much to do with the "raw mode" of a connected modem.

Are you sure you don't mix up things ?

**Reply by sethstenzel** [reply]
Thanks for your reply, I was searching for PPP I think, or PPPoS (PPP Over Serial). However, I can't seem to get any sort of PPP working.
I'm using a cellular modem, the Quectel BG95M3. It does have a mode for PPP, but I can't seem to authenticate or get a proper IP when using that implementation. I tried to enable some debugging with custom firmware but can't seem to make any progress or figure out how to resolve it.

Do you know if there is a way in firmware to debug out any data placed on or received by UART?

The Quectel modems have a UDP, TCP, and SSL mode, which you provide with the host and port, and in the case of SSL, and related certs, it then makes the socket connection, and you can then enter a raw mode passing data as you normally would with a socket connection.

I guess I'm looking for a way that I can use the modem native configurations which establish that socket connection and then pass that connection over in such a way that normal MicroPython libraries can use that connection as it would a normal network connection.

I do think PPPoS is what I'd ideally want, but I can't seem to get it working. Here you can see some of my code. Note that those are the AT commands to enter PPP mode as [referenced here](https://www.quectel.com/wp-content/uploads/2021/11/Quectel_BG95BG77BG600L_Series_PPP_Application_Note_V1.0.pdf).

```
def ppp_attempt(client):
    serial_link = client.modem.__uart
    client.cmd('AT+QICSGP=1,1,"data.apn.name","","",1')  # set mobile apn
    client.cmd('AT+CGDCONT=1,"IP","data.apn.name",,0,0')  # set apn
    client.cmd('AT+QIDNSGIP=1,"8.8.8.8"')  # set dns
    client.cmd("ATD*99#")  # activate ppp mode
    minutes = 60
    pppos = network.PPP(serial_link)
    while not pppos.isconnected() and minutes > 0:
        pppos.active(True)
        pppos.connect()
        attempts = 20
        while not pppos.isconnected() and attempts > 0:
            attempts -= 1
            status = pppos.status()
            print(pppos.ifconfig())
            connection_state = pppos.isconnected()
            print(f"[PPP] Connection state: {connection_state}. Status {status}")
            sleep(3)
        minutes -= 1
        pppos.active(False)

    if pppos.connected():
        print(
            "Connected",
            connection_state,
        )
        print(pppos.ifconfig())
        ping("8.8.8.8")
    sleep(1)
    client.cmd("+++")  # leave transparent ppp mode
    sleep(1)
```

This is the result when I try to use PPP
```
DATA:<['ATD*99#', 'CONNECT 150000000']>

netif: netmask of interface  set to 255.255.255.255
netif: added interface pp IP addr 0.0.0.0 netmask 255.255.255.255 gw 0.0.0.0
ppp phase changed[2]: phase=0
netif: setting default interface pp
ppp_connect[2]: holdoff=0
ppp phase changed[2]: phase=3
pppos_connect: unit 2: connecting
ppp_start[2]
ppp phase changed[2]: phase=6
pppos_send_config[2]: out_accm=FF FF FF FF
ppp_send_config[2]
pppos_recv_config[2]: in_accm=FF FF FF FF
ppp_recv_config[2]
ppp: auth protocols: PAP=0 CHAP=0 CHAP_MD5=0
sent [LCP ConfReq id=0x1 <asyncmap 0x0> <magic 0xe285233a> <pcomp> <accomp>]
pppos_write[2]: len=24
ppp_start[2]: finished
('0.0.0.0', '0.0.0.0', '255.255.255.255', '0.0.0.0')
[PPP] Connection state: False. Status None
pppos_input[2]: got 256 bytes
rcvd [LCP ConfReq id=0x0 <asyncmap 0x0> <auth pap> <magic 0x2f0bc11d> <pcomp> <accomp>]
No auth is possible
sent [LCP ConfRej id=0x0 <auth pap>]
pppos_write[2]: len=12
rcvd [LCP ConfAck id=0x1 <asyncmap 0x0> <magic 0xe285233a> <pcomp> <accomp>]
rcvd [LCP ConfReq id=0x1 <asyncmap 0x0> <auth pap> <magic 0x2f0bc11d> <pcomp> <accomp>]
No auth is possible
sent [LCP ConfRej id=0x1 <auth pap>]
pppos_write[2]: len=12
rcvd [LCP ConfReq id=0x2 <asyncmap 0x0> <auth pap> <magic 0x2f0bc11d> <pcomp> <accomp>]
No auth is possible
sent [LCP ConfRej id=0x2 <auth pap>]
pppos_write[2]: len=12
('0.0.0.0', '0.0.0.0', '255.255.255.255', '0.0.0.0')
[PPP] Connection state: False. Status None
```
